backend/server/public/src/admin.jsx

```diff
@@ -244,42 +237,93 @@ const Ad = ({ ad, onSuppressClick, onPermalinkClick }) => (
   </div>
 );
 
-let Ads = ({
-  ads,
-  onSuppressClick,
-  onKeyUp,
-  onPermalinkClick,
-  search,
-  pagination
-}) => (
-  <div id="ads">
-    <input
-      id="search"
-      placeholder="Search for ads"
-      onKeyUp={onKeyUp}
-      search={search}
-    />
-    {pagination ? <Pagination /> : ""}
-    {ads.map(ad => (
-      <Ad
-        ad={ad}
-        key={ad.id}
-        onSuppressClick={onSuppressClick}
-        onPermalinkClick={onPermalinkClick}
+class AdDetail extends React.Component {
+  constructor(props) {
+    super(props);
+  }
+
+  componentDidMount() {
+    console.log(this.props);
+    let ad_id = null;
+    if (this.props.match) {
+      ad_id = this.props.match.params.ad_id;
+    } else {
+      ad_id = this.props.requested_ad_id; // really state.permalinked_ad.requested_ad_id
+    }
+    store.dispatch(getOneAd(ad_id));
+  }
+
+  render() {
+    if (this.props.ads && this.props.ads[this.props.requested_ad_id].loaded) {
+      if (this.props.ads[this.props.requested_ad_id].id) {
+        return (<div id="ad">
+          <input id="search" placeholder="Search for ads" />
+
+          <Ad ad={this.props.ads[this.props.requested_ad_id]} onSuppressClick={this.props.onSuppressClick} />
+
+        </div>);
+      } else {
+        return (<div><h2>Uh oh, an ad with that ID couldn&apos;t be found!</h2></div>);
+      }
+    } else {
+      return (<div><h2>Loading...</h2></div>);
+    }
+  }
+}
+AdDetail = withRouter(connect(
+  ({ permalinked_ad }) => (
+    { // this is a mapStateToProps function. { ads } is destructuring the `store` hash and getting the `ads` element.
+      ads: permalinked_ad.ads,
+      requested_ad_id: permalinked_ad.requested_ad_id,
+
+    }),
+  (dispatch) => ({ // ownProps is available as a second argument here.
+    onSuppressClick: (ad) => dispatch(suppressAd(ad))
+  })
+)(AdDetail));
+
+class Ads extends React.Component {
+  constructor(props) {
+    super(props);
+  }
+
+  componentDidMount() {
+    deserialize(store.dispatch);
+    refresh(store).then(() => store.subscribe(() => refresh(store))); // anytime anything changes, then make the ajax request whenever the user changes the facets they want.
```

Comment:
This is a bug. If people go from permalink -> search -> permalink -> search we'll have multiple listeners subscribed. I think we need to make refresh an action which it isn't right now.